### ISSUE TITLE
Completion not considering context [JavaScript]

### ISSUE THREAD

**Issue by sQVe** [autozimu/LanguageClient-neovim#408]
_I'll like to first give a huge thank you for this awesome plugin. I've been using it extensively and it's awesome!_ :hugs:

## Summary

This is more a question than an issue. When using `javascript-typescript-langserver` together with `LanguageClient-neovim` there's seem to be some slight quirks regarding context. A fresh file with the content:
```
class Dummy {}

const array = [1,2,3];
array.<cursor>
```
In the above code the completion will actually suggest `Dummy`, which seems a bit off.

Is this a **bug** or in some way wanted behaviour?

## Reproduction
```
NVIM v0.2.2
languageclient 0.1.73
javascript-typescript-langserver 2.8.0
```
[LanguageClient.log](https://github.com/autozimu/LanguageClient-neovim/files/1959207/LanguageClient.log)
[LanguageServer.log](https://github.com/autozimu/LanguageClient-neovim/files/1959208/LanguageServer.log)

## Current Behavior
![screenfetch-2018-04-29_22-38-22](https://user-images.githubusercontent.com/2284724/39410891-86104440-4bff-11e8-889d-019c06d1899e.png)

## Expected Behavior

The completion should consider cursor placement and in this case only suggest things that make sense on a array after `.`. Like `map`, `includes` and etc.

**Comment by chemzqm** [comment]
It's issue of javascript-typescript-langserver, you won't have those unnecessary complete items when using typescript.

**Comment by autozimu** [comment]
Thanks for your kind words.

Yes, as prevuous comment, it is a lanuage server issue. Please report at the lanuage server repo.

**Comment by chemzqm** [comment]
@sQVe Actually, you can change this line https://github.com/autozimu/LanguageClient-neovim/blob/f2cad27cffe4677bcf33801310af15ade4cce6fc/rplugin/python3/deoplete/sources/LanguageClientSource.py#L15 to

```
        self.sorters = []
```
the methods would come first.

**Comment by sQVe** [comment]
@autozimu @chemzqm Cheers for the input. I'll take it the language server instead.

**Comment by sQVe** [comment]
@autozimu This issue is being thrown back and forth now. See https://github.com/sourcegraph/javascript-typescript-langserver/issues/451. I'm a bit unsure how to proceed - where should I raise this issue.

**Comment by autozimu** [comment]
```
13:32:27 INFO main src/vim.rs:90 => Some("javascript") {"jsonrpc":"2.0","method":"textDocument/completion","params":{"position":{"character":6,"line":3},"textDocument":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js"}},"id":24}
13:32:27 INFO reader-javascript src/vim.rs:364 <= Some("javascript") {"jsonrpc":"2.0","id":24,"result":{"isIncomplete":false,"items":[{"label":"length","kind":10,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"length"}},{"label":"toString","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"toString"}},{"label":"toLocaleString","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"toLocaleString"}},{"label":"push","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"push"}},{"label":"pop","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"pop"}},{"label":"concat","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"concat"}},{"label":"join","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"join"}},{"label":"reverse","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"reverse"}},{"label":"shift","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"shift"}},{"label":"slice","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"slice"}},{"label":"sort","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"sort"}},{"label":"splice","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"splice"}},{"label":"unshift","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"unshift"}},{"label":"indexOf","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"indexOf"}},{"label":"lastIndexOf","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"lastIndexOf"}},{"label":"every","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"every"}},{"label":"some","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"some"}},{"label":"forEach","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"forEach"}},{"label":"map","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"map"}},{"label":"filter","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"filter"}},{"label":"reduce","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"reduce"}},{"label":"reduceRight","kind":2,"sortText":"0","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"reduceRight"}},{"label":"Dummy","sortText":"1","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"Dummy"}},{"label":"array","sortText":"1","data":{"uri":"file:///Users/lijunfen/repo/LanguageClient-neovim/tests/data/sample-js/src/index.js","offset":45,"entryName":"array"}}]}}
```

This is extracted from my reproduction. As it shows, `Dummy` is included as the second last item.

The problem was show up in language server author's is because it was shown last in VSCode.
![image](https://user-images.githubusercontent.com/1453551/39449149-23e81482-4c7c-11e8-8a37-616e668f557f.png)

**Comment by sQVe** [comment]
@autozimu Then maybe the real issue here is more related to how this client actually sorts the results kontra the suggestions made by the langserver.

I tried @chemzqm's suggestion above and removed `sorter_word` and that made a **BIG** difference. The suggestions now pretty much mimic VSCode's behavior spot on. Why is the client concerned with the sorting of the suggestions? I'm guessing that `sorter_word` sorts in alphabetic order or something?

___________
_EDIT_

I've now tried it with both JavaScript & Haskell and would say I prefer the completions without `sorter_word`.
